src/kudu/consensus/log_cache.cc

```diff
@@ -164,17 +197,182 @@ void LogCache::TruncateOpsAfterUnlocked(int64_t index) {
+  buffer.resize(compressed_len);
+  VLOG(2) << "Compressed OpId: " <<  msg->get()->id().ShortDebugString() <<
+             " original payload size: " << uncompressed_slice.size() <<
+             " compressed payload size: " << compressed_len;
+
+  // Now create a new replicate message and copy contents from original message
+  // and compressed payload
+  std::unique_ptr<ReplicateMsg> rep_msg(new ReplicateMsg);
+  *(rep_msg->mutable_id()) = msg->get()->id();
+  rep_msg->set_timestamp(msg->get()->timestamp());
+  rep_msg->set_op_type(msg->get()->op_type());
+
+  WritePayloadPB* write_payload = rep_msg->mutable_write_payload();
+  write_payload->set_payload(std::move(buffer.ToString()));
+  write_payload->set_compression_codec(codec->type());
+  write_payload->set_uncompressed_size(payload_str.size());
+
+  compressed_msg->reset(rep_msg.release());
+  return Status::OK();
+}
+
 Status LogCache::AppendOperations(const vector<ReplicateRefPtr>& msgs,
                                   const StatusCallback& callback) {
   CHECK_GT(msgs.size(), 0);
 
   // SpaceUsed is relatively expensive, so do calculations outside the lock
   // and cache the result with each message.
   int64_t mem_required = 0;
+  int64_t total_msg_size = 0;
   vector<CacheEntry> entries_to_insert;
   entries_to_insert.reserve(msgs.size());
+
+  bool found_compressed_msgs = false;
   for (const auto& msg : msgs) {
-    CacheEntry e = { msg, static_cast<int64_t>(msg->get()->SpaceUsedLong()) };
+    CacheEntry e;
+    e.mem_usage = 0;
+    e.msg_size = msg->get()->SpaceUsedLong();
+
+    const auto op_type = msg->get()->op_type();
+    bool is_compressed =
+      (msg->get()->write_payload().compression_codec() != NO_COMPRESSION);
+
+    if (is_compressed) {
+      // This batch has a compressed message. Store this fact to be used later
+      // when appending this batch to the log
+      found_compressed_msgs = true;
+    }
+
+    // Try to compress the message if:
+    // (1) this msg is not already compressed (identified by 'is_compressed').
+    // This can happen when this AppendOperations() is called on a secondary and
+    // the primary has already compressed this msg
+    // (2) Compression is configured for this instance (codec_ should not be
+    // nullptr)
+    // (3) This is a 'WRITE_OP_EXT' type (config-change/no-op/rotates are not
+    // compressed today - these messages are small and can be left uncompressed)
+    const CompressionCodec* codec = codec_.load();
+    if (!is_compressed && codec && op_type == WRITE_OP_EXT) {
+      std::unique_ptr<ReplicateMsg> compressed_msg;
+      auto status =
+        CompressMsg(msg, log_cache_compression_buf_, &compressed_msg);
+      if (status.ok()) {
+        // Successfully compressed this msg. So, use the compressed msg to cache
+        e.mem_usage = static_cast<int64_t>(compressed_msg->SpaceUsedLong());
+        e.msg = make_scoped_refptr_replicate(compressed_msg.release());
+      }
```

Comment:
some kind of an error message under vlog or every 5 minutes would tell us that compression failed